File: Backend/src/controllers/auth.controller.js
import { supabase, supabaseAdmin } from '../config/supabase.js';

// POST /api/auth/register
export const register = async (req, res, next) => {
  try {
    const { name, email, password } = req.body;

    // Use admin client to create user — skips email confirmation in all environments
    const { data, error } = await supabaseAdmin.auth.admin.createUser({
      email,
      password,
      user_metadata: { name },
      email_confirm: true, // auto-confirm so user can log in immediately
    });

    if (error) return res.status(400).json({ message: error.message });

    res.status(201).json({
      message: 'Registration successful.',
      user: { id: data.user?.id, email: data.user?.email, name },
    });
  } catch (error) {
    next(error);
  }
};

// POST /api/auth/login
export const login = async (req, res, next) => {
  try {
    const { email, password } = req.body;

    const { data, error } = await supabase.auth.signInWithPassword({ email, password });

    if (error) return res.status(401).json({ message: error.message });

    res.json({
      token: data.session.access_token,
      refreshToken: data.session.refresh_token,
      user: {
        id: data.user.id,
        email: data.user.email,
        name: data.user.user_metadata?.name,
      },
    });
  } catch (error) {
    next(error);
  }
};

// POST /api/auth/logout
export const logout = async (req, res, next) => {
  try {
    const { error } = await supabase.auth.signOut();
    if (error) return res.status(400).json({ message: error.message });
    res.json({ message: 'Logged out successfully' });
  } catch (error) {
    next(error);
  }
};

// POST /api/auth/refresh
export const refreshToken = async (req, res, next) => {
  try {
    const { refresh_token } = req.body;
    if (!refresh_token) return res.status(400).json({ message: 'Refresh token required' });

    const { data, error } = await supabase.auth.refreshSession({ refresh_token });

    if (error) return res.status(401).json({ message: error.message });

    res.json({
      token: data.session.access_token,
      refreshToken: data.session.refresh_token,
    });
  } catch (error) {
    next(error);
  }
};

// GET /api/auth/me  (protected)
export const getMe = async (req, res, next) => {
  try {
    // Fetch profile from our public.profiles table
    const { data: profile, error } = await supabaseAdmin
      .from('profiles')
      .select('id, name, created_at')
      .eq('id', req.user.id)
      .single();

    if (error) return res.status(404).json({ message: 'Profile not found' });

    res.json({
      user: {
        id: req.user.id,
        email: req.user.email,
        name: profile.name,
        createdAt: profile.created_at,
      },
    });
  } catch (error) {
    next(error);
  }
};

File: FastAPI_Backend/app/controllers/auth_controller.py
from fastapi import HTTPException
from app.config.supabase import supabase, supabase_admin
from app.models.auth import (
    RegisterRequest, LoginRequest, RefreshRequest,
    RegisterResponse, LoginResponse, UserResponse,
)


async def register(body: RegisterRequest) -> RegisterResponse:
    try:
        res = supabase_admin.auth.admin.create_user({
            "email": body.email,
            "password": body.password,
            "user_metadata": {"name": body.name},
            "email_confirm": True,
        })
        if not res.user:
            raise HTTPException(status_code=400, detail="Registration failed")

        return RegisterResponse(
            message="Registration successful.",
            user=UserResponse(
                id=str(res.user.id),
                email=res.user.email or "",
                name=body.name,
            ),
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))


async def login(body: LoginRequest) -> LoginResponse:
    try:
        res = supabase.auth.sign_in_with_password({
            "email": body.email,
            "password": body.password,
        })
        if not res.session:
            raise HTTPException(status_code=401, detail="Invalid email or password")

        name = None
        if res.user and res.user.user_metadata:
            name = res.user.user_metadata.get("name")

        return LoginResponse(
            token=res.session.access_token,
            refreshToken=res.session.refresh_token,
            user=UserResponse(
                id=str(res.user.id),
                email=res.user.email or "",
                name=name,
            ),
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=401, detail=str(e))


async def logout() -> dict:
    try:
        supabase.auth.sign_out()
        return {"message": "Logged out successfully"}
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))


async def refresh_token(body: RefreshRequest) -> dict:
    try:
        res = supabase.auth.refresh_session(body.refresh_token)
        if not res.session:
            raise HTTPException(status_code=401, detail="Invalid refresh token")
        return {
            "token": res.session.access_token,
            "refreshToken": res.session.refresh_token,
        }
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=401, detail=str(e))


async def get_me(current_user) -> dict:
    try:
        res = supabase_admin.table("profiles") \
            .select("id, name, created_at") \
            .eq("id", str(current_user.id)) \
            .single() \
            .execute()

        if not res.data:
            raise HTTPException(status_code=404, detail="Profile not found")

        return {
            "user": {
                "id": str(current_user.id),
                "email": current_user.email,
                "name": res.data.get("name"),
                "createdAt": res.data.get("created_at"),
            }
        }
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))
